File: exercism/java/twelve-days/src/main/java/TwelveDays.java
class TwelveDays {

    private static final String ON_THE_XTH_DAY = "On the %s day of Christmas my true love gave to me: ";
    private static final String[] XTH = new String[]{"first", "second", "third", "fourth", "fifth", "sixth", "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth"};
    private static final String[] PRESENTS = new String[]{
            "a Partridge in a Pear Tree.\n",
            "two Turtle Doves, and ",
            "three French Hens, ",
            "four Calling Birds, ",
            "five Gold Rings, ",
            "six Geese-a-Laying, ",
            "seven Swans-a-Swimming, ",
            "eight Maids-a-Milking, ",
            "nine Ladies Dancing, ",
            "ten Lords-a-Leaping, ",
            "eleven Pipers Piping, ",
            "twelve Drummers Drumming, "
    };

    String verse(int verseNumber) {
        StringBuilder sb = new StringBuilder(String.format(ON_THE_XTH_DAY, XTH[verseNumber - 1]));
        for (int i = verseNumber - 1; i >= 0; i--) {
            sb.append(PRESENTS[i]);
        }
        return sb.toString();
    }

    String verses(int startVerse, int endVerse) {
        StringBuilder sb = new StringBuilder();
        for (int i = startVerse; i <= endVerse; i++) {
            sb.append(this.verse(i));
            if (i < endVerse) {
                sb.append("\n");
            }
        }
        return sb.toString();
    }

    String sing() {
        return this.verses(1, 12);
    }
}

File: exercism/python/twelve-days/twelve_days.py
def recite_verse(verse_number):
    """Function that the verse from the twelve days of Christmas specified by its number"""
    
    verse_start = "On the {} day of Christmas my true love gave to me: "
    days = ["first", "second", "third", "fourth", "fifth", "sixth",
            "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth"]
    gifts = [
        "a Partridge in a Pear Tree.",
        "two Turtle Doves,",
        "three French Hens, ",
        "four Calling Birds, ",
        "five Gold Rings, ",
        "six Geese-a-Laying, ",
        "seven Swans-a-Swimming, ",
        "eight Maids-a-Milking, ",
        "nine Ladies Dancing, ",
        "ten Lords-a-Leaping, ",
        "eleven Pipers Piping, ",
        "twelve Drummers Drumming, "
    ]

    verse = verse_start.format(days[verse_number-1])

    if verse_number == 1:
        verse = verse + gifts[0]
    else:
        for i in range(verse_number-1, 0, -1):
            verse = verse + gifts[i]
        verse = verse + " and " + gifts[0]

    return verse


def recite(start_verse, end_verse):
    """Function that gives as a list the verses from the twelve days of Christmas between start_verse and end_verse"""
    return [recite_verse(i) for i in range(start_verse, end_verse+1)]
